torchrl/objectives/ppo.py

```diff
@@ -764,16 +782,19 @@ def forward(self, tensordict: TensorDictBase) -> TensorDictBase:
         gain2 = log_weight_clip.exp() * advantage
 
         gain = torch.stack([gain1, gain2], -1).min(dim=-1)[0]
-        td_out = TensorDict({"loss_objective": -gain.mean()}, [])
+        td_out = TensorDict({"loss_objective": -gain}, batch_size=tensordict.batch_size)
 
         if self.entropy_bonus:
             entropy = self.get_entropy_bonus(dist)
-            td_out.set("entropy", entropy.mean().detach())  # for logging
-            td_out.set("loss_entropy", -self.entropy_coef * entropy.mean())
+            td_out.set("entropy", entropy.detach())  # for logging
+            td_out.set("loss_entropy", -self.entropy_coef * entropy)
         if self.critic_coef:
             loss_critic = self.loss_critic(tensordict)
-            td_out.set("loss_critic", loss_critic.mean())
-        td_out.set("ESS", ess.mean() / batch)
+            td_out.set("loss_critic", loss_critic)
+
+        if self.reduction is not None:
+            td_out = td_out.apply(lambda x: _reduce(x, self.reduction), batch_size=[])
+            td_out.set("ESS", _reduce(ess, self.reduction) / batch)
```

Comment:
it's a bit problematic that the output depends on the reduction... 
I'm not sure that having the batch-size set is going to work across all losses anyway (some losses subsample a part of data to compute part of the loss).
I agree it isn't super clear though but maybe the best we can do to keep things generic.

Reply:
What about adding all losses with the correct shape in the first level and add a nested TensorDict with all the other tensors. We can call it `metadata` as suggested in the first comment.

Reply:
that won't work because the nested tensordict must have the same leading dims as the root (+ other optional dims) as it is the case for tensors

Reply:
aah true...